wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -0,0 +1,260 @@
+                Ok(dxc_blob) => (
+                    Ok(crate::dx12::CompiledShader::Dxc(dxc_blob.to_vec())),
+                    log::Level::Info,
+                ),
+                Err(e) => (
+                    Err(crate::PipelineError::Linkage(
+                        stage_bit,
+                        format!("DXC compile error: {}", e),
+                    )),
+                    log::Level::Error,
+                ),
+            },
+            Err(e) => (
+                Err(crate::PipelineError::Linkage(
+                    stage_bit,
+                    format!("DXC compile error: {:?}", e),
+                )),
+                log::Level::Error,
+            ),
+        };
+
+        (result, log_level)
+    }
+
+    impl From<hassle_rs::HassleError> for crate::DeviceError {
+        fn from(value: hassle_rs::HassleError) -> Self {
+            match value {
+                hassle_rs::HassleError::Win32Error(e) => {
+                    // TODO: This returns an HRESULT, should we try and use the associated Windows error message?
+                    log::error!("Win32 error: {e:?}");
+                    crate::DeviceError::Lost
+                }
+                hassle_rs::HassleError::LoadLibraryError { filename, inner } => {
+                    log::error!("Failed to load dxc library {filename:?}. Inner error: {inner:?}");
+                    crate::DeviceError::Lost
+                }
+                hassle_rs::HassleError::LibLoadingError(e) => {
+                    log::error!("Failed to load dxc library. {e:?}");
+                    crate::DeviceError::Lost
+                }
+                hassle_rs::HassleError::WindowsOnly(e) => {
+                    log::error!("Signing with dxil.dll is only supported on Windows. {e:?}");
+                    crate::DeviceError::Lost
+                }
+                // `ValidationError` and `CompileError` should never happen in a context involving `DeviceError`
+                hassle_rs::HassleError::ValidationError(_e) => unimplemented!(),
+                hassle_rs::HassleError::CompileError(_e) => unimplemented!(),
+            }
+        }
+    }
+}
+
+// These are stubs for when the `dxc_shader_compiler` feature is disabled.
+#[cfg(not(feature = "dxc_shader_compiler"))]
+mod shader {
+    use std::path::PathBuf;
+
+    pub(crate) struct DxcContainer {}
+
+    pub(crate) fn get_dxc_container(
+        _dxc_path: Option<PathBuf>,
+        _dxil_path: Option<PathBuf>,
+    ) -> Result<Option<DxcContainer>, crate::DeviceError> {
+        // Falls back to Fxc and logs an error.
+        log::error!("DXC shader compiler was requested on Instance creation, but the DXC feature is disabled. Enable the `dxc_shader_compiler` feature on wgpu_hal to use DXC.");
+        Ok(None)
+    }
+
+    // It shouldn't be possible that this gets called with the `dxc_shader_compiler` feature disabled, but just in case...
+    pub(crate) fn compile_dxc(
```

Comment:
This only ever gets called when the DxcContainer field on `Device` is `Some<DxcContainer>`, but it will/should always be `None` if the dxc_shader_compiler feature is disabled. Not sure if it should be `unimplemented!()`, or leave it stubbed in as a safety net.